File: solutions/python/matching-brackets/1/matching_brackets.py
def is_paired(input_string):
    # Define matching pairs
    brackets = {'(': ')', '[': ']', '{': '}'}
    opening = set(brackets.keys())  # Opening brackets
    closing = set(brackets.values())  # Closing brackets
    
    stack = []
    
    for char in input_string:
        if char in opening:
            # Push opening bracket onto stack
            stack.append(char)
        elif char in closing:
            # Check if stack is empty (no matching opening bracket)
            if not stack:
                return False
            
            # Check if closing bracket matches the most recent opening bracket
            last_opening = stack.pop()
            if brackets[last_opening] != char:
                return False
    
    # If stack is empty, all brackets were properly matched
    return len(stack) == 0

File: solutions/javascript/matching-brackets/1/matching-brackets.js
//
// This is only a SKELETON file for the 'Matching Brackets' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const isPaired = (input) => {
  // Stack to keep track of opening brackets
  const stack = [];
  
  // Map of closing brackets to their corresponding opening brackets
  const bracketPairs = {
    ')': '(',
    ']': '[',
    '}': '{'
  };
  
  // Set of opening brackets for quick lookup
  const openingBrackets = new Set(['(', '[', '{']);
  
  // Process each character in the input
  for (const char of input) {
    // If it's an opening bracket, push it onto the stack
    if (openingBrackets.has(char)) {
      stack.push(char);
    } 
    // If it's a closing bracket, check if it matches
    else if (char in bracketPairs) {
      // If stack is empty or top of stack doesn't match, it's unbalanced
      if (stack.length === 0 || stack.pop() !== bracketPairs[char]) {
        return false;
      }
    }
    // Ignore all other characters
  }
  
  // If stack is empty, all brackets were properly matched
  return stack.length === 0;
};